### PR TITLE
[PowerPC] option `-msoft-float` should not block the PC-relative address instruction

### LINKED ISSUE #62372: Clang powerpc64le pcrel code model is not used if -msoft-float or similar flags are used
Example here - https://godbolt.org/z/1YhsEW7En

Remove `-msoft-float` or switch to GCC for expected pcrel code model. `-mno-altivec` and `-mno-vsx` flags at least also cause this. This prevents building the Linux kernel with the pcrel code model.

### PR DESCRIPTION
**Pull request by diggerlin**
The Prefix instruction is introduced on PowerPC ISA3_1.

In the PR,  
1. The `FeaturePrefixInstrs` do not imply the `FeatureP8Vector` ,`FeatureP9Vector` . 
2. `FeaturePrefixInstrs`  implies only the FeatureISA3_1.
3. For the prefix instructions `paddi` and `pli`  , they have  `Predicates = [PrefixInstrs] `
4. For the prefix instructions `plfs` and `plfd`,  they have  `Predicates = [PrefixInstrs, HasFPU] `
5. For the prefix instructions  "plxv` , "plxssp` and `plxsd` , they have `Predicates = [PrefixInstrs, HasP10Vector]`

Fixes #62372

### CODE REVIEW

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 1]
```diff
@@ -18258,9 +18258,11 @@ unsigned PPCTargetLowering::computeMOFlags(const SDNode *Parent, SDValue N,
     FlagSet |= PPC::MOF_SubtargetBeforeP9;
   else {
```
Comment (chenzheng1030) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:18259]: nit: the `{}` can be removed now?

File: llvm/lib/Target/PowerPC/PPCInstrP10.td [hunk 2]
```diff
@@ -1157,15 +1173,15 @@ let Predicates = [PairedVectorMemops] in {
...
+  let Predicates = [PairedVectorMemops, PrefixInstrs, HasP10Vector] in {
     def : Pat<(v256i1 (int_ppc_vsx_lxvp PDForm:$src)), (PLXVP memri34:$src)>;
   }
   // Intrinsics for Paired Vector Stores.
   def : Pat<(int_ppc_vsx_stxvp v256i1:$XSp, DQForm:$dst),
             (STXVP $XSp, memrix16:$dst)>;
   def : Pat<(int_ppc_vsx_stxvp v256i1:$XSp, XForm:$dst),
             (STXVPX $XSp, XForm:$dst)>;
-  let Predicates = [PairedVectorMemops, PrefixInstrs] in {
+  let Predicates = [PairedVectorMemops, PrefixInstrs, HasP10Vector] in {
```
Comment (chenzheng1030) [inline llvm/lib/Target/PowerPC/PPCInstrP10.td:1184]: There are some other def/pat with predicate `PrefixInstrs` unfixed? For example, at line 1338, 1364, 2145, 2282.

File: llvm/test/CodeGen/PowerPC/issue-62372-bug.ll [hunk 3]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc -o - %s| FileCheck %s
+
+target triple = "powerpc64le-unknown-linux-gnu"
+
+@bar = dso_local global i32 0, align 4
+
+define dso_local ptr @foo() #0 {
+entry:
+	  ret ptr @bar
```
Comment (chenzheng1030) [inline llvm/test/CodeGen/PowerPC/issue-62372-bug.ll:9]: nit, the format here

File: llvm/test/CodeGen/PowerPC/issue-62372-bug.ll [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+attributes #0 = { noinline nounwind optnone uwtable "frame-pointer"="all" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="pwr10" "target-features"="+altivec,+bpermd,+crbits,+crypto,+direct-move,+extdiv,+isa-v206-instructions,+isa-v207-instructions,+isa-v30-instructions,+isa-v31-instructions,+mma,+paired-vector-memops,+pcrelative-memops,+power10-vector,+power8-vector,+power9-vector,+prefix-instrs,+quadword-atomics,+vsx,-aix-small-local-dynamic-tls,-aix-small-local-exec-tls,-hard-float,-htm,-privileged,-rop-protect,-spe" "use-soft-float"="true" }
```
Comment (chenzheng1030) [inline llvm/test/CodeGen/PowerPC/issue-62372-bug.ll:12]: Is it possible to tune the attributes a little bit? Just use the ones that matter? I don't believe the `"frame-pointer"="all" "no-trapping-math"="true"` will impact the result.

File: llvm/test/CodeGen/PowerPC/issue-62372-bug.ll [hunk 5]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc -o - %s| FileCheck %s
```
Comment (chenzheng1030) [inline llvm/test/CodeGen/PowerPC/issue-62372-bug.ll:1]: nit: we usually use name pr62372.ll for a fix targeting for a github issue.

**Review by chenzheng1030 — COMMENTED** [review]
Thanks for improving this.
- In the description, `Fixes #62372` should be a right form to connect this pr and the issue.
- We may also need to check all the places where `hasPrefixInstrs()` is called. Now `hasPrefixInstrs()` does not necessary mean vector instructions are also enabled, so we need to add check for `hasP10Vector()` too. For example for the usages in
  - `llvm/lib/Target/PowerPC/PPCRegisterInfo.cpp`
  - `llvm/lib/Target/PowerPC/PPCISelLowering.cpp`

**Review by amy-kwan — COMMENTED** [review]
One more comment on the test

File: llvm/lib/Target/PowerPC/PPCInstrP10.td [hunk 6]
```diff
@@ -683,6 +702,28 @@ let Predicates = [PrefixInstrs] in {
...
+                                  "pstxssp $RST, $addr", "pstxssp $RST, $D", IIC_LdStLFD>;
+    defm PSTXSD :
+      8LS_DForm_R_SI34_RTA5_MEM_p<46, (outs), (ins vfrc:$RST, (memri34 $D, $RA):$addr),
+                                  (ins vfrc:$RST, (memri34_pcrel $D, $RA):$addr),
+                                  (ins vfrc:$RST, s34imm_pcrel:$D),
+                                  "pstxsd $RST, $addr", "pstxsd $RST, $D", IIC_LdStLFD>;
+  }
+}
+
+let Predicates = [PrefixInstrs] in {
```
Comment (chenzheng1030) [inline llvm/lib/Target/PowerPC/PPCInstrP10.td:725]: Maybe we can add a follow up patch to group the instruction/pattern defs by the Predicates. At least `[PrefixInstrs]` and `[PrefixInstrs, HasP10Vector]` are used in more than 1 place. We can do this later.

File: llvm/test/CodeGen/PowerPC/pr92543.ll [hunk 7]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc -ppc-asm-full-reg-names -o - %s| FileCheck %s
```
Comment (chenzheng1030) [inline llvm/test/CodeGen/PowerPC/pr92543.ll:1]: the file name should contain the bug issue number 62372, not the patch number.(a little confusing : ))

**Review by chenzheng1030 — COMMENTED** [review]
overall looks good.

Some comments about simplify the test case and a possible follow up.

File: llvm/lib/Target/PowerPC/PPCInstrP10.td (hunk 6 above)
Comment (diggerlin) [reply llvm/lib/Target/PowerPC/PPCInstrP10.td:725]: I will create a NFC patch for it later.

File: llvm/test/CodeGen/PowerPC/pr62372.ll [hunk 8]
```diff
@@ -0,0 +1,12 @@
+; RUN: llc -ppc-asm-full-reg-names -mcpu=pwr10 -mtriple powerpc64le-unknown-linux-gnu -o - %s| FileCheck %s
```
Comment (amy-kwan) [inline llvm/test/CodeGen/PowerPC/pr62372.ll:1]: nit: Line too long. Can we split up the line?

**Review by chenzheng1030 — APPROVED** [review]
LGTM as long as @amy-kwan 's comment is addressed.

Please wait for some days for other reviewers.

Thanks very much for the improvements.

**Review by amy-kwan — APPROVED** [review]
LGTM, thanks Digger!

**Pull request merged**